add hidden columns support

khoomeister opened the pull request:
Todo

* [x] Bump version number
* [ ] After merge, tag & push

albertywu commented on slick.grid.js:
```diff
@@ -684,7 +690,9 @@ if (typeof Slick === "undefined") {
 
       // Get needed data for this column
       var columnDef = columns[columnIndex];
-      newEl = options.subHeaderRenderers[rowIndex](columnDef);
```
nice catch with the `var` here...

khoomeister replied:
thank `'use strict'` for that :smiley:

albertywu commented on slick.grid.js:
```diff
@@ -3749,12 +3782,80 @@ if (typeof Slick === "undefined") {
+      delete(column.showHidden);
+    }
+
+    function showHiddenColumnsByDirection(column, columnDirection) {
```
it's not immediately obvious what `columnDirection` is. Can you drop a comment to help the API consumer here?

khoomeister replied:
Thanks, I've added it!

albertywu commented on slick.grid.js:
```diff
@@ -1522,23 +1540,31 @@ if (typeof Slick === "undefined") {
+        columnIdxById[m.id] = i;
+
         // Changing the object reference can cause problems for external consumers of that object, so we're careful to maintain it using this crazy double extend.
-        tempCol = $.extend({}, columnDefaults, m);
```
another nice catch... wow... makes me wonder what other bugs are lurking here

bcherny commented on slick.grid.css:
```diff
@@ -30,6 +30,9 @@
 .slickGrid {
   overflow: hidden;
   outline: 0; }
+  .slickGrid .hidden {
```
can we call it isHidden?

bcherny commented on slick.grid.js:
```diff
@@ -3749,12 +3782,80 @@ if (typeof Slick === "undefined") {
+    }
+
+    function isHiddenColumnVisible(column) {
+      return column.hidden && column.showHidden;
```
can you fill me in on what the difference between these 2 states is?

khoomeister replied:
So a hidden column can be temporarily made visible by user interaction, but it's not persisted.

Hence there's this subtle difference between hidden/unhidden columns and visible/invisible columns.

Does that make sense?

bcherny replied:
got it. thanks!

albertywu commented on slick.grid.js:
```diff
@@ -1441,18 +1455,22 @@ if (typeof Slick === "undefined") {
-        canvas = i > options.pinnedColumn ? contentCanvas[1].width : contentCanvas[0].width;
-        rule.right.style.right = (canvas - x - w) + "px";
+        var column = columns[i];
+        var width = getColumnVisibleWidth(column);
```
The best practice i've seen is to put all var declarations at top-of-scope (so how it was before). I like it because I can rely on seeing all locally-scoped variable declarations in one place.

Is there a good reason for doing it this way?

khoomeister replied:
I generally like `var`s closer to where they're used.  Would you prefer me to stick to all `var`s at the top?

bcherny replied:
im with chris on this, though we do have the opposite convention in slickgrid.*.js

albertywu replied:
OK, let's keep as-is then.

albertywu commented:
:+1:

khoomeister commented:
BTW, I've refactored `showAdjacentHiddenColumns` to `iterateColumnsInDirection` so I could DRY up the newly added `getNextVisibleColumn` as well.

Also added jsig to all touched & new functions

bcherny commented:
dont forget to bump the version! :+1:
